scipy/sparse/csgraph/tests/test_matching.py

```diff
@@ -143,3 +144,98 @@ def test_large_random_graph_with_one_edge_incident_to_each_vertex():
     # Should get identity matrix back
     assert_equal(any(C1.diagonal() == 0), False)
     assert_equal(any(C2.diagonal() == 0), False)
+
+
+def test_min_weight_full_matching_empty_graph():
+    graph = csr_matrix((0, 0))
+    row_ind, col_ind = min_weight_full_bipartite_matching(graph)
+    assert len(row_ind) == 0
+    assert len(col_ind) == 0
+
+
+def test_min_weight_full_matching_empty_left_partition():
+    graph = csr_matrix((2, 0))
+    row_ind, col_ind = min_weight_full_bipartite_matching(graph)
+    assert len(row_ind) == 0
+    assert len(col_ind) == 0
+
+
+def test_min_weight_full_matching_empty_right_partition():
+    graph = csr_matrix((0, 3))
+    row_ind, col_ind = min_weight_full_bipartite_matching(graph)
+    assert len(row_ind) == 0
+    assert len(col_ind) == 0
+
+
+def test_min_weight_full_matching_infeasible_square():
+    graph = csr_matrix([[1, 1, 1], [1, 0, 0], [1, 0, 0]])
+    with pytest.raises(ValueError):
+        min_weight_full_bipartite_matching(graph)
+
+
+def test_min_weight_full_matching_infeasible_other_square():
+    graph = csr_matrix([[1, 1, 1], [0, 0, 1], [0, 0, 1]])
+    with pytest.raises(ValueError):
+        min_weight_full_bipartite_matching(graph)
+
+
+def test_min_weight_full_matching_infeasible_rectangular():
+    graph = csr_matrix([[1, 0, 0], [2, 0, 0]])
+    with pytest.raises(ValueError):
+        min_weight_full_bipartite_matching(graph)
+
+
+def test_min_weight_full_matching_infeasible_other_rectangular():
+    graph = csr_matrix([[1, 0], [2, 0], [5, 0]])
+    with pytest.raises(ValueError):
+        min_weight_full_bipartite_matching(graph)
+
+
+def test_min_weight_full_matching_for_various_inputs():
+    for sign in [-1, 1]:
+        for graph, expected_cost in [
+            ([[400, 150, 400],
+              [400, 450, 600],
+              [300, 225, 300]],
+             [150, 400, 300]
+             ),
+
+            ([[400, 150, 400, 1],
+              [400, 450, 600, 2],
+              [300, 225, 300, 3]],
+             [150, 2, 300]),
+
+            ([[10, 10, 8],
+              [9, 8, 1],
+              [9, 7, 4]],
+             [10, 1, 7]),
+
+            ([[10, 10, 8, 11],
+              [9, 8, 1, 1],
+              [9, 7, 4, 10]],
+             [10, 1, 4]),
+
+            ([[10, 0, 0],
```

Comment:
Does this algorithm handle infinities in the cost matrix? Would it be worth duplicating this test with the zeros replaced with `float("inf")`?

Reply:
Thanks for working your way through these!

Regarding the handling of infinities, that's actually the main reason for why one might want something like this in the first place: Whenever there's a `float('inf')` in what you've set up with `scipy.optimize.linear_sum_assignment`, there might as well not be an edge in the first place. The entries missing in the sparse biadjacency matrix are exactly those that correspond to edges not present in the graph. That is, even if it does perform well enough in general, one might want to use `min_weight_full_bipartite_matching` *in particular* when there are many infinities.

Now, SciPy convention has it that in the constructors of sparse matrices, zero entries are exactly the ones that are removed when converting a dense matrix to its sparse counterpart. And that's why there are zeros in the test: I simply took your test and replaced all `float('inf')` with zeros.

Now there is one slight risk of confusion here (certainly if you didn't realize immediately, that suggests that we can do better here): We now have no way of representing edges with weight 0 in the algorithm a priori. I didn't worry too much about that here, since this issue isn't unique to this PR and affects all parts of `sparse.csgraph`; much care needs to be taken in handling explicit zeros if those are to be used. In practice one can of course always shift all weights away from zero, but that still means wrestling an implementation detail. FWIW, I've made explicit note of this in the example section of the documentation.

Reply:
> Whenever there's a float('inf') in what you've set up with scipy.optimize.linear_sum_assignment, there might as well not be an edge in the first place

Yes - I completely agree. I have checked out your branch and tested that it already handles infinities properly, so it is just a question of whether it is worth adding an extra test. I think it is, since, eventually, some user is going to put a `float("inf")` in the cost matrix and call MWFBM.

Reply:
Ah, sorry, now I see what you're asking! Sorry for the redundant elaboration. :) I haven't considered that case; I believe it does work as well (the algorithm is very close to the one you implemented), but I would have to run through once more to be sure. I'm wondering, though, if it would be simpler to modify the inputs and simply remove any infinite entry entirely. It might wreck performance in some situations though.

Reply:
Went for the route of adjusting the inputs. Logic being that the conversions that happen during validation can cause explicit zeros to vanish, so the only consistent approach is to ensure that they never appear. And if we do so, we may as well get rid of the (positive) infinities in the process.

I updated the documentation to spell out this behavior as explicitly as possible and included the "missing" test.